### ISSUE TITLE
Redis VM not working on Solaris

### ISSUE THREAD

**Issue by GoogleCodeExporter** [ashish21590/redis#237]
```
What steps will reproduce the problem?

compile with:
CC=gcc DEBUG= gmake 

configure with:
vm-enabled yes
 (and other settings low enough to fit into swapfs)

start with my vm conf:
./redis-server redis.vm.conf

What is the expected output? What do you see instead?

Expect it to proceed, but it hangs after this:

[22412] 08 May 22:45:35 * Using '/tmp/redis-swap.vm' as swap file
[22412] 08 May 22:45:35 * Allocating 402653184 bytes of swap file
[22412] 08 May 22:45:35 * Swap file allocated with success
[22412] 08 May 22:45:35 - Allocated 1572864 bytes page table for 12582912 pages

Nothing about the server starting.

What version of the product are you using? On what operating system?
Solaris "11" snv_121 (basically the tip of the public branch)
Tested with HEAD of github, currently 585af7e, but also with problem appearing 
earlier.

It works well in the in-memory version, but I'm not able to go beyond that.
```

Original issue reported on code.google.com by `a.lind...@gmail.com` on 8 May 2010 at 10:56

**Comment by GoogleCodeExporter** [comment]
```
I have the same problem i think.

I'm on:
SunOS xxx 5.10 Generic_141414-07 sun4u sparc SUNW,Sun-Fire-V245

and i tested with the 2.0.0-RC1

xxx:~/redis-2.0.0-rc1$ ./redis-server ./redis.vm.conf
[19393] 24 May 10:36:51 * Using '/tmp/redis.swap' as swap file
[19393] 24 May 10:36:51 * Allocating 4294967296 bytes of swap file
[19393] 24 May 10:36:51 * Swap file allocated with success
^C
xxx:~/redis-2.0.0-rc1$ ./redis-server ./redis.conf
[19401] 24 May 10:37:19 * Server started, Redis version 1.3.14
[19401] 24 May 10:37:19 * DB loaded from disk: 0 seconds
[19401] 24 May 10:37:19 * The server is now ready to accept connections on port 
6379
^C

the one with the vm just spins with 100% cpu - I waited about 10 minutes before 
i
gave up.

Is there anything I can do to help debug the problem?

```

Original comment by `anko.com` on 24 May 2010 at 12:39

**Comment by GoogleCodeExporter** [comment]
```
Please can someone check with GDB where it starts looping?

Thanks,
Salvatore
```

Original comment by `anti...@gmail.com` on 31 Aug 2010 at 10:56

**Comment by GoogleCodeExporter** [comment]
Original comment by `anti...@gmail.com` on 31 Aug 2010 at 10:56
- Changed state: **Started**

**Comment by GoogleCodeExporter** [comment]
```
8905        pthread_attr_setstacksize(&server.io_threads_attr, stacksize);
seems to be where it freezes
```

Original comment by `anko.com` on 3 Sep 2010 at 12:19

**Comment by GoogleCodeExporter** [comment]
```
i should add the debug level log messages before this are
[8800] 03 Sep 10:22:38 * Using '/tmp/redis.swap' as swap file
[8800] 03 Sep 10:22:38 * Allocating 4294967296 bytes of swap file
[8800] 03 Sep 10:22:38 * Swap file allocated with success
[8800] 03 Sep 10:22:38 - Allocated 16777216 bytes page table for 134217728 pages

```

Original comment by `anko.com` on 3 Sep 2010 at 12:23

**Comment by GoogleCodeExporter** [comment]
```
actually I lie! 
turns out the problem is that 
pthread_attr_getstacksize(&server.io_threads_attr, &stacksize);

is returning a stacksize of 0, so the next line;

while (stacksize < REDIS_THREAD_STACK_SIZE) stacksize *= 2;

never does anything!

I fixed it by putting in
if (!stacksize) {
    stacksize = 1;
}

before

while (stacksize < REDIS_THREAD_STACK_SIZE) stacksize *= 2;


```

Original comment by `anko.com` on 3 Sep 2010 at 3:59

**Comment by GoogleCodeExporter** [comment]
```
Cool thanks! Just merged in Redis master :)

This fix will be backported to 2.0.0 as well.

Cheers,
Salvatore
```

Original comment by `anti...@gmail.com` on 3 Sep 2010 at 8:24
- Changed state: **Verified**
